### ISSUE TITLE
[FR] Vendor distutils stubs

### ISSUE THREAD

**Issue by Avasam** [pypa/setuptools#4689]
### What's the problem this feature will solve?

Type-checkers don't see distutils as available in Python 3.12+
This is already causing us much pain when it comes to type-checking, and is becoming a blocker.
One of the main issue is that importing from distutils, types are seen as `Any`, which breaks subclassing, removes tons of type-safety, and creates nasty inconsistencies when trying to type-check <3.12 vs >=3.12.

We also get some overloading issues when type-checkers don't understand we're using an updated distutils version vendored from pypa/distutils.

As long as `_distutils_hack` exists, we can't have type-checkers scan the vendored version instead. (if pypa/distutils was installed like a regular package, this wouldn't be an issue either)

### Describe the solution you'd like

My suggestion is to vendor `distutils-stubs`, reflecting the types as the modernized pypa/distutils.

Those stubs should exist in the repo in a folder named `distutils` or `distutils-stubs`.
`distutils-stubs` should be installed in users' site-packages when installing `setuptools` (this can be locked behind an extra if you have concerns for vendors)

### Alternative Solutions

- Make pypa/distutils a real package found in site-packages
- Change how the hack works, and always import from `._distutils` in setuptools instead.

### Additional context

- https://github.com/pypa/setuptools/pull/4504 (see failures for Python 3.12 and 3.13)
- https://github.com/pypa/setuptools/issues/2345#issuecomment-1627561833

### Code of Conduct

- [X] I agree to follow the PSF Code of Conduct

**Comment by abravalheri** [comment]
Thank you very much @Avasam for having a look on this.

I will leave the final call on this to Jason.

But my first impression is that this is a bit too intrusive to be handled in this repo. (We don't even maintain `distutils`, so it would not make much sense to maintain `distutils-stubs`, if anywhere, `pypa/distutils` would be a best place for them).

Now the approach in https://github.com/pypa/setuptools/pull/4704 is slightly better, but it still introduces a lot of noise and some level of extra maintenance. It would be nice if `mypy` instead had a configuration that allowed us to point where the `distutils` package should be loaded from, to workaround mypy general limitation regarding meta path finders.

I wonder if there would be another approach...

Let's assume that we all are willing to move `setuptools/_distutils` to `setuptools/_private/distutils` or something similar. In this case, would simply adding the configuration `mypy_path = $MYPY_CONFIG_FILE_DIR/setuptools/_private` enough to solve the problem? (i.e. would `mypy` accept the `setuptools/_private/distutils` directory as a source of truth for the typings in `distutils`, or we would still need to add those `.pyi` files?).

(Moving `setuptools/_distutils` to a different directory is still something that we have to discuss and I don't know if Jason would be OK with that, but it is certainly better than adding 50-something extra `.pyi` files).

**Comment by Avasam** [comment]
I agree that in the form of #4691, setuptools isn't the place to host these stubs. It is however, the best place for a PR to show the effects. (which probably means I should mark it as draft).
If going with that solution, I agree that it should live somewhere else (preferably with pypa/distutils, if it doesn't introduce distribution issues).

---

#4704 is different though, as setuptools is responsible for the distutils hack. From a separation of concerns pov, pypa/distutils shouldn't need to be aware that setuptools implements it using import hacks nor the exact folder name it lives in. In other words, setuptools is responsible for the distutils hack, and it should be responsible for the stub aliasing `distutils` to `setuptools._distutils` (or w/e folder location).

In the hypothetical that `setuptools/_distutils` is renamed to `setuptools/_private/distutils`, yes it means that locally for development, we no longer need a stub. *But* users of setuptools will still need one. And that's true as long as the distutils hack exists and as long as setuptools can't point to distutils as a regular dependency (so once Python 3.11 support is dropped ?). One advantage of the folder rename, is that the alias stubs could be generated on source install/build and included in wheels, so #4704 could potentially be updated to not have the stubs be part of the git source (it already includes the script to generate them)

---

In all cases, setuptools should be responsible to ship `stubs-distutils`, so that it gets installed (and maybe uninstalled) with it, even if it's only though an extra.

**Comment by Avasam** [comment]
https://github.com/pypa/setuptools/pull/4861 is now in a really good state and can be reviewed.

It does make use of a custom build backend for editable installs. But Maybe you or @jaraco can point me to a better implementation for that detail.

**Comment by jaraco** [comment]
> As long as `_distutils_hack` exists, we can't have type-checkers scan the vendored version instead.

I do very much want to eliminate this hack (https://github.com/pypa/setuptools/issues/3625). 



> (if pypa/distutils was installed like a regular package, this wouldn't be an issue either)

I've also considered this possibility, but the plan instead is to fully move distutils code into setuptools (and eliminate support for `import distutils`). One of the main reasons to avoid creating a separate package is we don't want to encourage users to install it separately.



> [#4861](https://github.com/pypa/setuptools/pull/4861) is now in a really good state and can be reviewed.

Thanks for this. On one hand, I'm really reluctant to adopt a specialized backend (and the interactions that has with the rest of the build) to address this issue. I don't want to introduce complications that will create flux for integrators or impede the migration off distutils. On the other hand, if it's temporary, maybe it doesn't matter how invasive it is.

I won't oppose the proposal, but I'd prefer to focus on getting distutils into a single location.

**Comment by Avasam** [comment]
@jaraco 
It's been a while, but I think I solely added the custom backend to support editable installs ?
W/o that it could be confusing for a contributor *who's developing with setuptools installed in editable mode* if there's errors in their IDE that don't match what the ` tox run -- -k mypy` command.

But as long as the `tox` command works, maybe you'd be fine with that and I can update my PR to not use a custom build backend. (of course I'd have to check that was indeed the only reason I added it, and not due to some issue with how the `tox` test installs the project)

---

> As long as _distutils_hack exists, we can't have type-checkers scan the vendored version instead.

Indeed for consumers of setuptools, but for the project itself that's not quite true:
As previously discussed (likely lost in some conversation). If `distutils` was vendored in setuptools in a folder actually called `distutils`. We could point mypy/pyright to the parent of that folder as a "source of types".
Meaning neither the `tox` command nor editable installs would need to generate stubs. In fact you wouldn't even need to install setuptools (generating the stubs) at all to typecheck the project (since the reference to `distutils` source types will be static).

You'd still need to generate the stubs like I'm doing for packaging/publishing though.
